### PR TITLE
Display events functions of extensions in a tree view

### PR DESCRIPTION
**Pull request by D8H**
https://gdevelop-storybook.s3.amazonaws.com/function-tree/latest/index.html?path=/story/eventsfunctionsextensioneditor-index--default

### CODE REVIEW

**Comment by AlexandreSi** [comment]
Feedback using the story:
- the drop indicator gives the impression you can drop a behavior inside another one. TreeViewRow might need a props to configure this. Or you might just need to change the `canDrop` callback
- the copy paste feature does not seem to work, is it because it's not implemented in the story? Or maybe it's just this UX path that does not work: I copy a method from a behavior that I want to paste inside an empty behavior => the context menu item "Paste" is greyed out.

File: newIDE/app/src/UI/EditorMosaic/index.js [hunk 1]
```diff
@@ -157,6 +157,58 @@ const removeNode = (
...
+    };
+  }
+  return {
+    ...currentNode,
+    first: resizeNode(currentNode.first, resizedNode, splitPercentage),
+    second: resizeNode(currentNode.second, resizedNode, splitPercentage),
+  };
+};
+
+const getNodeSize = (
```
Comment (AlexandreSi) [inline newIDE/app/src/UI/EditorMosaic/index.js:190]: No need to test these 2 methods?

File: newIDE/app/src/EventsFunctionsList/index.js [hunk 2]
```diff
@@ -6,509 +6,1291 @@ import { t } from '@lingui/macro';
...
-      >
-        <AsyncIcon fontSize="small" style={styles.tooltip} />
-      </Tooltip>
-      {label}
-    </>
-  ) : (
-    label
-  );
-};
+class ObjectTreeViewItem implements TreeViewItem {
```
Comment (AlexandreSi) [inline newIDE/app/src/EventsFunctionsList/index.js:113]: Should it be named `ExtensionTreeViewItem`?

Comment (AlexandreSi) [reply newIDE/app/src/EventsFunctionsList/index.js:113]: Oh ok, it's for a CustomObject in fact

Comment (D8H) [reply newIDE/app/src/EventsFunctionsList/index.js:113]: It's for event-based objects.

File: newIDE/app/src/EventsBasedBehaviorEditor/EventsBasedBehaviorEditorDialog.js
Comment (AlexandreSi) [inline newIDE/app/src/EventsBasedBehaviorEditor/EventsBasedBehaviorEditorDialog.js:1]: Rename file to remove Dialog from it?

File: newIDE/app/src/EventsBasedObjectEditor/EventsBasedObjectEditorDialog.js
Comment (AlexandreSi) [inline newIDE/app/src/EventsBasedObjectEditor/EventsBasedObjectEditorDialog.js:1]: Same - rename file?

File: newIDE/app/src/EventsFunctionsList/FunctionTreeViewItemContent.js [hunk 3]
```diff
@@ -0,0 +1,441 @@
...
+        <AsyncIcon fontSize="small" style={styles.tooltip} />
+      </Tooltip>
+    ) : null;
+    return privateIcon && asyncIcon ? (
+      <>
+        {privateIcon}
+        {asyncIcon}
+      </>
+    ) : (
+      privateIcon || asyncIcon
```
Comment (AlexandreSi) [inline newIDE/app/src/EventsFunctionsList/FunctionTreeViewItemContent.js:290]: It feels a bit complicated. You can use a list for this:
```
const leftIcons = []
if (...) {
  leftIcons.push(...)
}
if (...) {
  leftIcons.push(...)
}
return leftIcons.length > 0 ? leftIcons : null
```
And make sure you add the key props to the array children.

File: newIDE/app/src/EventsFunctionsList/FunctionTreeViewItemContent.js [hunk 4]
```diff
@@ -0,0 +1,441 @@
...
+
+  _deleteEventsFunction = (
+    eventsFunction: gdEventsFunction,
+    { askForConfirmation }: {| askForConfirmation: boolean |}
+  ) => {
+    const { eventsFunctionsContainer } = this.props;
+
+    if (askForConfirmation) {
+      const answer = Window.showConfirmDialog(
+        "Are you sure you want to remove this function? This can't be undone."
```
Comment (AlexandreSi) [inline newIDE/app/src/EventsFunctionsList/FunctionTreeViewItemContent.js:318]: Can you translate this? If it's too complicated, don't bother.

File: newIDE/app/src/EventsFunctionsList/index.js [hunk 5]
```diff
@@ -6,509 +6,1291 @@ import { t } from '@lingui/macro';
...
+    const addNewEventsFunction = React.useCallback(
+      (itemContent: ?TreeViewItemContent) => {
+        const eventsFunctionsContainer = itemContent
+          ? itemContent.getEventsFunctionsContainer() ||
+            eventsFunctionsExtension
+          : eventsFunctionsExtension;
+
+        const index = itemContent && itemContent.getFunctionInsertionIndex();
+
+        // Let EventsFunctionsExtensionEditor knows if the function is:
```
Comment (AlexandreSi) [inline newIDE/app/src/EventsFunctionsList/index.js:614]:
```suggestion
        // Let EventsFunctionsExtensionEditor know if the function is:
```

File: newIDE/app/src/EventsFunctionsList/index.js [hunk 6]
```diff
@@ -6,509 +6,1291 @@ import { t } from '@lingui/macro';
...
+        editName,
+        eventsFunctionsExtension,
+        forceUpdate,
+        scrollToItem,
+        setSelectionToObject,
+        unsavedChanges,
+      ]
+    );
+
+    // TODO
```
Comment (AlexandreSi) [inline newIDE/app/src/EventsFunctionsList/index.js:913]: What should be done?

File: newIDE/app/src/EventsFunctionsList/BehaviorTreeViewItemContent.js [hunk 7]
```diff
@@ -0,0 +1,335 @@
...
+const EVENTS_BASED_BEHAVIOR_CLIPBOARD_KIND = 'Events Based Behavior';
+
+const styles = {
+  tooltip: { marginRight: 5, verticalAlign: 'bottom' },
+};
+
+export const getBehaviorTreeViewItemId = (
+  eventsBasedBehavior: gdEventsBasedBehavior
+): string => {
+  return 'behaviors.' + eventsBasedBehavior.getName();
```
Comment (AlexandreSi) [inline newIDE/app/src/EventsFunctionsList/BehaviorTreeViewItemContent.js:32]: Using the name as id means if you rename the behavior, it closes :/

File: newIDE/app/src/EventsFunctionsList/index.js [hunk 8]
```diff
@@ -6,509 +6,1211 @@ import { t } from '@lingui/macro';
...
-  render() {
-    const {
-      project,
-      eventsFunctionsContainer,
-      selectedEventsFunction,
-      onSelectEventsFunction,
-    } = this.props;
-    const { searchText } = this.state;
+    const canMoveSelectionTo = React.useCallback(
+      (destinationItem: TreeViewItem, where: 'before' | 'inside' | 'after') =>
```
Comment (D8H) [inline newIDE/app/src/EventsFunctionsList/index.js:1002]: I didn't used the `TreeViewItemContent` for this function but it may be easier to read as it would split the logic by targeted item type. I guess it's small enough to avoid to split it.

File: newIDE/app/src/EventsFunctionsList/index.js [hunk 9]
```diff
@@ -6,509 +6,1211 @@ import { t } from '@lingui/macro';
...
+      (itemContent: ?TreeViewItemContent) => {
+        const eventsFunctionsContainer = itemContent
+          ? itemContent.getEventsFunctionsContainer() ||
+            eventsFunctionsExtension
+          : eventsFunctionsExtension;
 
-    const newName = newNameGenerator(name, name =>
-      eventsFunctionsContainer.hasEventsFunctionNamed(name)
-    );
+        const index = itemContent && itemContent.getFunctionInsertionIndex();
```
Comment (D8H) [inline newIDE/app/src/EventsFunctionsList/index.js:545]: `getFunctionInsertionIndex` and `getEventsFunctionsContainer` are only used here. It doesn't feel like a good encapsulation. I wonder if there should rather be an `addFunction` passed in props to the `TreeViewItemContent` which would give the container and index to this function. But it would add an other callback level and may make the call tree hard to follow.

File: newIDE/app/src/EventsFunctionsList/index.js [hunk 10]
```diff
@@ -6,509 +6,1211 @@ import { t } from '@lingui/macro';
...
       </Background>
     );
   }
-}
+);
+
+const arePropsEqual = (prevProps: Props, nextProps: Props): boolean =>
+  // The component is costly to render, so avoid any re-rendering as much
+  // as possible.
+  // We make the assumption that no changes to objects list is made outside
```
Comment (4ian) [inline newIDE/app/src/EventsFunctionsList/index.js:1191]:
```suggestion
  // We make the assumption that no changes to the extension is made outside
```

File: newIDE/app/src/EventsFunctionsList/index.js (hunk 8 above)
Comment (4ian) [reply newIDE/app/src/EventsFunctionsList/index.js:1002]: Polymorphism would tell us to move this into a method, but yeah it's ok as it's fairly small I guess! :)

File: newIDE/app/src/EventsFunctionsList/ObjectTreeViewItemContent.js [hunk 11]
```diff
@@ -0,0 +1,296 @@
...
+    eventsBasedObject: gdEventsBasedObject,
+    newName: string,
+    cb: (boolean) => void
+  ) => void,
+  onEventsBasedObjectRenamed: (eventsBasedObject: gdEventsBasedObject) => void,
+|};
+
+export type EventObjectProps = {|
+  ...TreeItemProps,
+  ...EventObjectCallbacks,
```
Comment (4ian) [inline newIDE/app/src/EventsFunctionsList/ObjectTreeViewItemContent.js:39]: I know it's verbose, but I think it's worth avoiding "EventObject" and keeping "EventsBasedObject" everywhere. If anything, this helps to grep the codebase for it.

```suggestion
export type EventsBasedObjectCallbacks = {|
  onSelectEventsBasedObject: (eventsBasedObject: ?gdEventsBasedObject) => void,
  onDeleteEventsBasedObject: (
    eventsBasedObject: gdEventsBasedObject,
    cb: (boolean) => void
  ) => void,
  onRenameEventsBasedObject: (
    eventsBasedObject: gdEventsBasedObject,
    newName: string,
    cb: (boolean) => void
  ) => void,
  onEventsBasedObjectRenamed: (eventsBasedObject: gdEventsBasedObject) => void,
|};

export type EventsBasedObjectProps = {|
  ...TreeItemProps,
  ...EventsBasedObjectCallbacks,
```

File: newIDE/app/src/EventsFunctionsList/FunctionTreeViewItemContent.js [hunk 12]
```diff
@@ -0,0 +1,428 @@
...
+const styles = {
+  tooltip: { marginRight: 5, verticalAlign: 'bottom' },
+};
+
+export type EventsFunctionCreationParameters = {|
+  functionType: 0 | 1 | 2,
+  name: ?string,
+|};
+
+export type EventFunctionCallbacks = {|
```
Comment (4ian) [inline newIDE/app/src/EventsFunctionsList/FunctionTreeViewItemContent.js:40]:
```suggestion
export type EventsFunctionCallbacks = {|
```

File: newIDE/app/src/EventsFunctionsList/index.js [hunk 13]
```diff
@@ -6,509 +6,1211 @@ import { t } from '@lingui/macro';
...
+        const eventFunctionProps = {
+          eventsBasedObject,
+          eventsFunctionsContainer,
+          ...eventFunctionCommonProps,
+        };
+        setSelectedItems([
+          new LeafTreeViewItem(
+            new FunctionTreeViewItemContent(eventsFunction, eventFunctionProps)
+          ),
+        ]);
```
[the remark this answers is not in the document]
Comment (4ian) [reply newIDE/app/src/EventsFunctionsList/index.js:524]: It indeed sounds a bit weird, when a selection is set, why a new item would be constructed?
Could we indeed query the tree to get the existing item associated to the function?

File: newIDE/app/src/EventsFunctionsList/index.js [hunk 14]
```diff
@@ -6,509 +6,1211 @@ import { t } from '@lingui/macro';
...
+    const eventBasedBehaviors = eventsFunctionsExtension.getEventsBasedBehaviors();
 
-  _duplicateEventsFunction = (
-    eventsFunction: gdEventsFunction,
-    newFunctionIndex: number
-  ) => {
-    const { eventsFunctionsContainer } = this.props;
-    const newName = newNameGenerator(eventsFunction.getName(), name =>
-      eventsFunctionsContainer.hasEventsFunctionNamed(name)
+    const eventBehaviorProps = React.useMemo<EventBehaviorProps>(
```
Comment (4ian) [inline newIDE/app/src/EventsFunctionsList/index.js:647]:
```suggestion
    const eventsBasedBehaviorProps = React.useMemo<EventBehaviorProps>(
```

File: newIDE/app/src/EventsFunctionsList/index.js [hunk 15]
```diff
@@ -6,509 +6,1211 @@ import { t } from '@lingui/macro';
...
+        scrollToItem,
+        showDeleteConfirmation,
+        unsavedChanges,
+      ]
     );
-    const newEventsFunction = eventsFunctionsContainer.insertEventsFunction(
-      eventsFunction,
-      newFunctionIndex
+
+    const eventObjectProps = React.useMemo<EventObjectProps>(
```
Comment (4ian) [inline newIDE/app/src/EventsFunctionsList/index.js:686]:
```suggestion
    const eventsBasedObjectProps = React.useMemo<EventsBasedObjectProps>(
```

File: newIDE/app/src/EventsFunctionsList/index.js [hunk 16]
```diff
@@ -6,509 +6,1211 @@ import { t } from '@lingui/macro';
...
+            eventObjectProps,
+            eventFunctionCommonProps
+          ),
+        ]);
+        onSelectEventsBasedObject(eventsBasedObject);
+      },
+      [eventFunctionCommonProps, eventObjectProps, onSelectEventsBasedObject]
+    );
+
+    const addNewEventsObject = React.useCallback(
```
Comment (4ian) [inline newIDE/app/src/EventsFunctionsList/index.js:803]:
```suggestion
    const addNewEventsBasedObject = React.useCallback(
```

**Review by 4ian — APPROVED** [review]
Very nice. Looks good to go to me.

**Comment by 4ian** [comment]
My main two concerns are :

- Would it be possible to avoid re-constructing items in setSelectionToBehavior (and other methods like it).
- Is this abstraction TreeViewItem something that should be backported to the tree view, and the existing objects editor adapted to use it?
